shared-libs/cht-datasource/src/libs/context.ts

```diff
@@ -0,0 +1,68 @@
+import { Doc } from './doc';
+import { hasField, isRecord } from './core';
+
+export type SourceDatabases = Readonly<{ medic: PouchDB.Database<Doc> }>;
+
+export interface SettingsService {
+  getAll: () => Doc;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface DataContext {}
+
+export interface LocalDataContext extends DataContext {
+  medicDb: PouchDB.Database<Doc>;
+  settings: SettingsService;
+}
+
+const assertSettingsService: (settings: unknown) => asserts settings is SettingsService = (settings: unknown) => {
+  if (!isRecord(settings) || !hasField(settings, {name: 'getAll', type: 'function'})) {
+    throw new Error(`Invalid settings service [${JSON.stringify(settings)}].`);
+  }
+};
+
+const assertSourceDatabases: (sourceDatabases: unknown) => asserts sourceDatabases is SourceDatabases =
+  (sourceDatabases: unknown) => {
+    if (!isRecord(sourceDatabases) || !hasField(sourceDatabases, {name: 'medic', type: 'object'})) {
+      throw new Error(`Invalid source databases [${JSON.stringify(sourceDatabases)}].`);
+    }
+  };
```

Comment:
Just so I am clear on what you are worried about here, is it around the janky `isRecord` and `hasField` stuff?  I feel like the general idea of these `asserts sourceDatabases is SourceDatabases` is solid since it fills a dual purpose.  It both allows for nudging the TS type compiler when needed, but more importantly provides a semantically proper way to do some schema validation for arguments coming in at the edge.  Let me know if you are seeing things differently here though!

Then, regarding the `isRecord` and `hasField` stuff, that just goes back to eventually [offloading schema validation](https://github.com/medic/cht-core/pull/9090#discussion_r1602262068) to a library...

Reply:
Yes my concern is that the validation is kind of fragile but there is no way to validate a function or an object "identity" in JS :man_shrugging: 
Let's keep it that way and if we come up with a better approach along the way we can always come back and change this code